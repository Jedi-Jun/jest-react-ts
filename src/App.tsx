import styled from 'styled-components';
import { Button, Input, ToDoItem } from 'components';
import { useState } from 'react';

const Container = styled.div`
  border: 1px solid red;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;
const Contents = styled.div`
  border: 1px solid rgba(0, 0, 0, 0.2);
  display: flex;
  flex-direction: column;
  background: #ffffff;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
  position: relative;
`;
const InputContainer = styled.div`
  display: flex;
`;
const Hr = styled.hr`
  border: none;
  width: 98%;
  height: 2px;
  background: #dbdbdb;
`;
const TodoListContainer = styled.div`
  // ...
`;

function App() {
  const [todo, setTodo] = useState('');
  const [todoList, setTodoList] = useState<string[]>([]);

  const handleAddTodo = (): void => {
    if (todo) {
      setTodoList([...todoList, todo]);
      setTodo('');
    }
  };
  const handleDeleteTodo = (index: number): void => {
    const list = [...todoList];
    list.splice(index, 1);
    setTodoList(list);
  };

  return (
    <Container>
      <Contents data-testid="foo">
        <TodoListContainer data-testid="todoList">
          {todoList.map((todo, index) => (
            <ToDoItem key={index} label={todo} onDelete={() => handleDeleteTodo(index)} />
          ))}
        </TodoListContainer>
        <Hr />
        <InputContainer>
          <Input
            value={todo}
            onChange={(text) => setTodo(text)}
            onKeyPress={(e) => (e.key === 'Enter' ? handleAddTodo() : null)}
            placeholder="Add something..."
          />
          <Button
            onClick={handleAddTodo}
            label="Add"
            backgroundColor="#3a86ff"
            hoverColor="#2877ff"
          />
        </InputContainer>
        <Button label="Del" backgroundColor="#ff1744" hoverColor="#d01440" />
      </Contents>
    </Container>
  );
}

export default App;
