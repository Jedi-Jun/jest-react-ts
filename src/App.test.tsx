import { render, screen, fireEvent } from '@testing-library/react';
import App from 'App';
import 'jest-styled-components';

describe('<App />', () => {
  // #1
  it('renders component correctly', () => {
    const { container } = render(<App />);
    const todoList = screen.getByTestId('todoList');
    expect(todoList).toBeInTheDocument();
    expect(todoList.firstChild).toBeNull();

    const input = screen.getByPlaceholderText('Add something...');
    expect(input).toBeInTheDocument();

    const label = screen.getByText('Add');
    expect(label).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
  // #2
  it('add and delete Todo items', () => {
    render(<App />);
    const input = screen.getByPlaceholderText('Add something...');
    const button = screen.getByText('Add');
    fireEvent.change(input, { target: { value: 'study Jest 1' } });
    fireEvent.click(button);

    const todoItem1 = screen.getByText('study Jest 1');
    expect(todoItem1).toBeInTheDocument();
    const deleteButton = screen.getByText('Del');
    expect(deleteButton).toBeInTheDocument();

    const todoList = screen.getByTestId('todoList');
    expect(todoList.childElementCount).toBe(1);

    fireEvent.change(input, { target: { value: 'study Jest 2' } });
    fireEvent.click(button);

    const todoItem2 = screen.getByText('study Jest 2');
    expect(todoItem2).toBeInTheDocument();
    expect(todoList.childElementCount).toBe(2);

    const deleteButtons = screen.getAllByText('Del');
    /* fireEvent.click(deleteButtons[0]);

    expect(todoItem1).not.toBeInTheDocument();
    expect(todoList.childElementCount).toBe(1); */
  });
});
