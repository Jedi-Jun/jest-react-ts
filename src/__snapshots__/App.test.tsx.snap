// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders component correctly 1`] = `
.c5 {
  background: #3a86ff;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
  border: 0;
}

.c5:hover {
  background: #2877ff;
}

.c5:active {
  box-shadow: inset 10px 10px 20px rgba(0,0,0,0.2);
}

.c7 {
  background: #ff1744;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
  border: 0;
}

.c7:hover {
  background: #d01440;
}

.c7:active {
  box-shadow: inset 10px 10px 20px rgba(0,0,0,0.2);
}

.c6 {
  color: #ffffff;
  font-size: 16px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 16px;
  padding: 8px 10px;
  border-radius: 8px;
  border: 3px solid #bdbdbd;
  outline: none;
}

.c0 {
  border: 1px solid red;
  min-height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  border: 1px solid rgba(0,0,0,0.2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: #ffffff;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  position: relative;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  border: none;
  width: 98%;
  height: 2px;
  background: #dbdbdb;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class=""
        data-testid="todoList"
      />
      <hr
        class="c2"
      />
      <div
        class="c3"
      >
        <input
          class="c4"
          placeholder="Add something..."
          type="text"
          value=""
        />
        <button
          class="c5"
        >
          <div
            class="c6"
          >
            Add
          </div>
        </button>
      </div>
      <button
        class="c7"
      >
        <div
          class="c6"
        >
          Remove
        </div>
      </button>
    </div>
  </div>
</div>
`;
